### PR TITLE
fix k8s overlays for addons

### LINKED ISSUE #21995: ServiceSpec is missing possibility to set annotations
**Describe the feature request**
I'd like to have in k8s.io.api.core.v1.ServiceSpec a possibility to add annotations to the service.

**Describe alternatives you've considered**
ATM there is no possibility to set annotations for the service

[ X] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Additional context**

### PR DESCRIPTION
**Pull request by richardwxn**
fix https://github.com/istio/istio/issues/21995
1. update the k8s overlay logic for addonComponents.
2. update kiali charts to include the service type and annotation options

Notes: the reordering of some fields in the test output is caused by the json merge generates different order with original yaml

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
I thought this is what the overlays were for?

File: operator/pkg/apis/istio/v1alpha1/values_types.proto [hunk 1]
```diff
@@ -2235,7 +2235,7 @@ message Values {
 
   TracingConfig tracing = 14;
 
-  KialiConfig kiali = 15;
+  TypeMapStringInterface kiali = 15;
```
Comment (howardjohn) [inline operator/pkg/apis/istio/v1alpha1/values_types.proto]: Why are we dropping all validation?

Comment (richardwxn) [reply operator/pkg/apis/istio/v1alpha1/values_types.proto]: introduced during testing, removed

File: operator/pkg/translate/translate.go [hunk 2]
```diff
@@ -147,9 +148,27 @@ func (t *Translator) OverlayK8sSettings(yml string, iop *v1alpha1.IstioOperatorS
 		}
 		inPath = strings.Replace(inPath, "gressGateways.", "gressGateways."+fmt.Sprint(index)+".", 1)
 		scope.Debugf("Checking for path %s in IstioOperatorSpec", inPath)
-		m, found, err := tpath.GetFromStructPath(iop, inPath)
-		if err != nil {
-			return "", err
+
+		var found bool
+		var m interface{}
+		if componentName == name.AddonComponentName {
```
Comment (ostromart) [inline operator/pkg/translate/translate.go]: it would be better to do this inside renderFeatureComponentPathTemplate as a special case template. this kind of fixup if going to be hard to read and maintain.

File: operator/pkg/translate/translate.go [hunk 3]
```diff
@@ -541,15 +560,23 @@ func renderFeatureComponentPathTemplate(tmpl string, componentName name.Componen
 
 // renderResourceComponentPathTemplate renders a template of the form <path>{{.ResourceName}}<path>{{.ContainerName}}<path> with
 // the supplied parameters.
-func (t *Translator) renderResourceComponentPathTemplate(tmpl string, componentName name.ComponentName) (string, error) {
+func (t *Translator) renderResourceComponentPathTemplate(tmpl string, componentName name.ComponentName, addonName string) (string, error) {
+	cn := string(componentName)
+	if componentName == name.AddonComponentName {
```
Comment (ostromart) [inline operator/pkg/translate/translate.go]: this should be commented. it will be hard to remember why this was done.

File: manifests/istio-telemetry/kiali/templates/service.yaml [hunk 4]
```diff
@@ -3,10 +3,15 @@ kind: Service
 metadata:
   name: kiali
   namespace: {{ .Release.Namespace }}
+  annotations:
```
Comment (ostromart) [inline manifests/istio-telemetry/kiali/templates/service.yaml]: per john's comment this strikes me as the wrong approach - we should not be adding things that are in IOP to values.yaml.

Comment (richardwxn) [reply manifests/istio-telemetry/kiali/templates/service.yaml]: But I think for now we should keep it consistent with other templates, which means that we should have it before we fully remove helm support.

Comment (ostromart) [reply manifests/istio-telemetry/kiali/templates/service.yaml]: we are quite likely to move the kiali chart out of the repo shortly so this is not likely to be a lasting change. does this pr work without this change? if not we will have to tackle this problem sooner or later.

Comment (richardwxn) [reply manifests/istio-telemetry/kiali/templates/service.yaml]: without this, user can still do overlay via IOP API, they just cannot use the helm values pass through, but we actually can do that for other addons. I think we'd better to remove all of them at once to keep it consistent later

File: operator/pkg/translate/translate.go (hunk 3 above)
Comment (richardwxn) [reply operator/pkg/translate/translate.go]: ok done

File: operator/pkg/translate/translate.go (hunk 2 above)
Comment (richardwxn) [reply operator/pkg/translate/translate.go]: I can create a helper function for this, would that be better? The reason is that we are not just changing the templates but there is also some logic changed required(due to the limitation of the tpath code we discussed offline)

**Comment by richardwxn** [comment]
@ostromart PTAL again

File: operator/cmd/mesh/manifest-generate_test.go [hunk 5]
```diff
@@ -303,7 +303,7 @@ func runTestGroup(t *testing.T, tests testGroup) {
 				}
 			}
 
-			if refreshGoldenFiles() {
+			if true {
```
Comment (ostromart) [inline operator/cmd/mesh/manifest-generate_test.go]: this doesn't look right - is it debug?

File: operator/cmd/mesh/testdata/manifest-generate/input/ingressgateway_k8s_settings.yaml [hunk 6]
```diff
@@ -23,5 +23,7 @@ spec:
           - name: foo
             port: 1234
             targetPort: 1234
+        serviceAnnotations:
+          manifest-generate: "testserviceAnnotation"
```
Comment (ostromart) [inline operator/cmd/mesh/testdata/manifest-generate/input/ingressgateway_k8s_settings.yaml]: we don't seem to be testing k8s overlays for addonComponents, so hard to tell if this change is working.

File: operator/cmd/mesh/manifest-generate_test.go (hunk 5 above)
Comment (richardwxn) [reply operator/cmd/mesh/manifest-generate_test.go]: oh yes good catch, it is for debug only

File: operator/cmd/mesh/testdata/manifest-generate/input/ingressgateway_k8s_settings.yaml (hunk 6 above)
Comment (richardwxn) [reply operator/cmd/mesh/testdata/manifest-generate/input/ingressgateway_k8s_settings.yaml]: done, added test

File: operator/pkg/translate/translate.go [hunk 7]
```diff
@@ -546,22 +547,55 @@ func renderFeatureComponentPathTemplate(tmpl string, componentName name.Componen
...
-		ContainerName: t.ComponentMaps[componentName].ContainerName,
+		ContainerName: cmp.ContainerName,
 	}
 	return util.RenderTemplate(tmpl, ts)
 }
 
+// getK8SSpecFromIOP is helper function to get k8s spec from IOP using path
+// 1. if component is not an addonComponent, get the spec directly from IOP
+// 2. otherwise convert the path and the root to point it to the entry of addonComponents map.
+// e.x: path "Components.AddonComponents.K8S.ReplicaCount" and root "Spec" ->
```
Comment (ostromart) [inline operator/pkg/translate/translate.go]: very hard to follow this example, could you rewrite in terms of API inputs and outputs?

File: operator/cmd/mesh/testdata/manifest-generate/input/addon_k8s_override.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+      enabled: true
+      k8s:
+        service:
+          externalTrafficPolicy: Local
+        serviceAnnotations:
+          test.annotation: test
+    kiali:
+      enabled: true
+      k8s:
+        serviceAnnotations:
```
[the remark this answers is not in the document]
Comment (richardwxn) [reply operator/cmd/mesh/testdata/manifest-generate/input/addon_k8s_override.yaml]: ok done

**Review by ostromart — APPROVED** [review]
